dbms/src/DataStreams/copyData.cpp

```diff
@@ -51,6 +54,74 @@ void copyDataImpl(IBlockInputStream & from, IBlockOutputStream & to, TCancelCall
 
 inline void doNothing(const Block &) {}
 
+
+void copyDataImpl(BlockInputStreams & froms, BlockOutputStreams & tos)
+{
+    if (froms.size() == tos.size())
+    {
+        std::vector<ThreadFromGlobalPool> threads;
+        threads.reserve(froms.size());
+        for (size_t i = 0; i < froms.size(); i++)
+        {
+            threads.emplace_back([from = froms.at(i), to = tos.at(i)]()
+            {
+                from->readPrefix();
+                to->writePrefix();
+                while (Block block = from->read())
+                    to->write(block);
+                from->readSuffix();
+                to->writeSuffix();
+            });
+        }
+        for (auto & thread : threads)
+            thread.join();
+    }
+    else
+    {
+        ConcurrentBoundedQueue<Block> queue(froms.size());
+        ThreadFromGlobalPool from_threads([&]()
+        {
+            std::vector<ThreadFromGlobalPool> from_threads_;
+            from_threads_.reserve(froms.size());
+            for (auto & from : froms)
+            {
+                from_threads_.emplace_back([&queue, from]()
+                {
+                    from->readPrefix();
+                    while (Block block = from->read())
+                        queue.push(block);
+                    from->readSuffix();
+                });
+            }
+            for (auto & thread : from_threads_)
+                thread.join();
+            for (size_t i = 0; i < tos.size(); i++)
+                queue.push({});
+        });
+        std::vector<ThreadFromGlobalPool> to_threads;
+        for (auto & to : tos)
+        {
+            to_threads.emplace_back([&queue, to]()
+            {
+                to->writePrefix();
+                Block block;
+                while (true)
+                {
+                    queue.pop(block);
```

Comment:
If there is an exception in reader, we will never get the empty block.
So, there will be a deadlock here.

Reply:
We push the empty blocks outside the reader threads (code line: 116), so I think we can ensure empty blocks are pushed even there is an exception in reader.